security/application.md

```diff
@@ -0,0 +1,555 @@
+up in TLS, JWT, and one-time passwords.
+
+### Passwords
+
+Note that for passwords, the attacker does not need to know the user's
+password _per se_; the attacker needs to know a string which will generate the
+desired hash. This is known as a collision attack.
+
+A rainbow table attack is done with a rainbow table: a giant list of every
+possible string and its resulting hash. Using such a list, the attacker can
+quickly look up the password given a hash.
+
+A similar attack is to, given one hash, run through every possible string,
+hashing each one, until you find a match.
+
+Rainbow table-style attacks have been on the rise since the early 1990s, making
+typical secure hashing functions inappropriate for passwords.
+
+The first solution is to use a salt: generate a random number, add that to the
+user's password, and hash _that_ string. Store the salt alongside the hashed
+password; each user gets their own salt.
+
+Salts destroy rainbow tables and cause headaches for hashing each string one at
+a time. But not enough of a headache: GPUs are at a point now where they can
+run secure hashing functions quickly. Too quickly.
+
+The solution is to use a key derivation algorithm. These are much like normal
+hashing algorithms (they're actually quite different, but that difference is
+negligible), except they are intentionally slow.
+
+The most common password hashing algorithms are bcrypt, scrypt, and PBKDF2.
+Each of these require a salt, but handle it themselves: the output of these
+functions is a string that contains the salt plus the hashed value. Store that
+entire string as the hashed password.
+
+## Encryption
+
+An encryption algorithm is one where a string can be made illegible and then
+returned back to the original string again, and where decrypting requires an
+out-of-band secret.
+
+Less abstractly: a string can be encrypted, and then to decrypt you must know
+the password.
+
+There are two kinds of encryption algorithms: symmetric and asymmetric.
+A symmetric algorithm is one where the same secret is used to encrypt it and
+decrypt it. An asymmetric algorithm is one where the string can be encrypted
+and decrypted using different secrets -- where the person encrypting cannot
+necessarily decrypt it.
+
+The most popular symmetric algorithms you'll encounter are AES and Twofish.
+These might be useful for encrypting a file to share with a group of people or
+for encrypting your filesystem. 1Password uses AES to encrypt an entire vault;
+it is encrypted at rest, and only decrypted when you enter the passphrase.
+
+Asymmetric encryption algorithms, also known as public/private keypair
+encryption, are more well-known -- in large part for how tricky they are to get
+right. Some famous ones are SSH, TLS (previously SSL), and PGP.  These start by
+generating a pair of encryption secrets known as the public and private keys.
+Anyone with the public key can encrypt a string, but only the holder of the
+private key can decrypt.
```

Comment:
Not sure if you want to get into this here, but it might be worth discussing how asymmetric encryption greatly increases the size of the thing you're encrypting while symmetric does not -- at least in the context of why databases don't use asymmetric encryption for encrypted fields.